recommendation_engine/autoencoder/train/train.py

```diff
@@ -0,0 +1,154 @@
+        hidden = latent_inputs
+        for i, hidden_dim in enumerate(self.hidden_dim[::-1], 1):  # Reverse because decoder.
+            hidden = Dense(hidden_dim, activation='sigmoid', name='hidden_d_{}'.format(i))(hidden)
+            logger.debug("Hooked up hidden layer with %d neurons" % hidden_dim)
+        # if hidden == latent_inputs:
+        #     logger.warning("No Hidden layers hooked up.")
+        outputs = Dense(original_dim, activation='sigmoid')(hidden)
+        decoder = Model(latent_inputs, outputs, name='decoder')
+
+        # Build the CVAE auto-encoder
+        outputs = decoder(encoder(inputs)[2])
+        cvae_model = Model(inputs, outputs, name='cvae')
+        # Load the pre-trained weights.
+        self.load_pretrain_weights(cvae_model)
+        reconstruction_loss = binary_crossentropy(inputs, outputs) * original_dim
+        kl_loss = 1 + z_log_sigma - tf.square(z_mean) - tf.exp(z_log_sigma)
+        kl_loss = -0.5 * tf.reduce_sum(kl_loss, axis=-1)
+
+        cvae_model.add_loss(tf.reduce_mean(reconstruction_loss + kl_loss))
+        cvae_model.compile(optimizer='adam')
+        cvae_model.summary()
+
+        # First load the weights from the pre-training
+        if self.pretrain_weights:
+            cvae_model = self.load_pretrain_weights(cvae_model)
+
+        saver = ModelCheckpoint(
+                check_path(TEMPORARY_CVAE_PATH),
+                save_weights_only=True,
+                verbose=1
+        )
+        tensorboard_config = TensorBoard(log_dir=check_path(TEMPORARY_CVAE_PATH))
+        # train the auto-encoder
+        cvae_model.fit(data, epochs=params_training.num_epochs,
+                       batch_size=params_training.batch_size,
+                       callbacks=[saver, tensorboard_config])
+        return self.encoder_z_mean
+
+
+if __name__ == '__main__':
+    x_train = np.load(os.path.join(TEMPORARY_PATH, 'content_matrix.npy'))
+    input_dim = x_train.shape[1]
+    p = TrainNetwork()
+    logger.info("Preprocessing of data started.")
+    p.get_preprocess_data.preprocess_data()
+
+    logger.info("size of training file is: ".format(len(x_train), len(x_train[0])))
+    user_to_item_matrix = load_rating(USER_ITEM_FILEPATH)
+    item_to_user_matrix = load_rating(ITEM_USER_FILEPATH)
+    logger.info("Shape of User and Item matrices:".format(np.shape(user_to_item_matrix),
+                                                          np.shape(item_to_user_matrix)))
+    pretrain.fit(x_train)
+    encoder_weights = p.train(x_train)
+    logger.info("Shape of encoder weights are: ".format(tf.shape(encoder_weights),
+                                                        len(encoder_weights),
+                                                        len(encoder_weights[0])))
+    pmf_obj = PMFTraining(len(user_to_item_matrix), len(item_to_user_matrix), encoder_weights)
+    logger.debug("PMF model has been initialised")
+    pmf_obj(user_to_item_matrix=user_to_item_matrix,
```

Comment:
How is this actually getting trained? Is this new to Tensorflow 2.0?

Reply:
Yeah, we ar using __call__ function. So after initialisation the object, It will automtically call.

Reply:
I know but how does it get trained inside the call? The `__call__` just defines the graph. Without `session.run()` how do you actually run iterations on the graph?

Reply:
Yeah Tensorflow 2.0 doesn't support session.run(), it uses tf.function() only.

Reply:
__call__ function makes the object callable so that it would work as:
`class_obj.__call__(arg1, arg2)`. So maybe earlier we used this for defining a graph and all.